Sprint 16

DmitryBeton opened the pull request:
- Экран со списком категорий реализован на архитектуре MVVM с имплементацией байндингов через замыкания.
- Сделаны экраны онбординга с использованием UIPageViewController.

DmitryBeton committed:
feat(category): Созданы экраны выбора и создания категории

DmitryBeton committed:
no message

DmitryBeton committed:
feat: Добавлен рабочий Onboarding экран & Функционал создания и выбора категории

DmitryBeton committed:
refactor(category): перевел category & createCategory на MVVM

DmitryBeton commented:
UILabel & UIButton разместил в OnboardingViewController тк не понял как иначе задать отступ в 24 единицы от button для pageControl.

DmitryBeton committed:
refactor: прожал ctrl i во всех файлах

DmitryBeton committed:
fix(category): Исправлен цвет текста кнопок при темной теме

Helvetii951 commented on Tracker/Modules/Category/CategoryView.swift:
```diff
@@ -0,0 +1,206 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell: UITableViewCell
+        
+        if let reusedCell = tableView.dequeueReusableCell(withIdentifier: "cell") {
```
Стоит поменять строку. Не совсем понятно что это за "cell" :)

Helvetii951 commented on Tracker/Modules/Category/CategoryView.swift:
```diff
@@ -0,0 +1,206 @@
+            cell.layer.maskedCorners = [.layerMinXMaxYCorner, .layerMaxXMaxYCorner]
+        } else {
+            cell.layer.cornerRadius = 0
+        }
```
Часто `if-else` предпочитают `switch`:
```swift

switch indexPath.row {
case 0:
    cell.layer.cornerRadius = 16
    cell.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
case viewModel.numberOfRows() - 1:
    cell.layer.cornerRadius = 16
    cell.layer.maskedCorners = [.layerMinXMaxYCorner, .layerMaxXMaxYCorner]
default:
    cell.layer.cornerRadius = 0
}
```

Helvetii951 commented on Tracker/Modules/Onboarding/OnboardingManager.swift:
```diff
@@ -0,0 +1,34 @@
+
+import Foundation
+
+final class OnboardingManager {
+    static let shared = OnboardingManager()
+    
+    private let userDefaults = UserDefaults.standard
+    private let hasSeenOnboardingKey = "hasSeenOnboarding"
+    
+    var hasSeenOnboarding: Bool {
```
👍

Helvetii951 commented on Tracker/Modules/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    
+    private let viewOne: UIViewController = OnboardingViewOne()
+    private let viewTwo: UIViewController = OnboardingViewTwo()
+    
```
Код `OnboardingViewOne` и `OnboardingViewTwo` идентичен. Можно создать один общий контроллер и передавать в него нужные данные:

```swift
struct PageModel {
    let image : UIImage?
    let text : String
}

extension PageModel {
    static let aboutTracking = PageModel(image: UIImage(named: "BlueBackground"),
                                        text: "Отслеживайте только то, что хотите")
    static let aboutWaterAndYoga = PageModel(image: UIImage(named: "BlueBackground"),
                                         text: "Даже если это не литры воды и йога")
}
...

let secondPageVC = OnboardingPageView(pageModel : .aboutWaterAndYoga)
```

Helvetii951 commented on Tracker/Modules/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        UIView.transition(with: window, duration: 0.3, options: .transitionCrossDissolve, animations: {
+            window.rootViewController = tabBarController
+        }, completion: nil)
+    }
```
Лучше забрать у контроллера онбординга обязанность навигации и отдать. ее тому, кто создал контроллер(в нашем случае это `SceneDelegate`). Передаем событие нажатия кнопки через, например, замыкание, а `SceneDelegate` уже пусть сам решает что делать.

Helvetii951 commented on Tracker/Modules/Category/CategoryView.swift:
```diff
@@ -0,0 +1,206 @@
+final class CategoryView: UIViewController {
+    
+    // MARK: - Properties
+    private let viewModel: CategoryViewModel
```
Для гибкости и тестируемости кода желательно объявлять вью-модель протоколом.

Helvetii951 commented on Tracker/Modules/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+            label.text = "Отслеживайте только то, что хотите"
+        } else {
+            label.text = "Даже если это не литры воды и йога"
+        }
```
```swift
        label.text = pageControl.currentPage == 0
        ? "Отслеживайте только то, что хотите"
        : "Даже если это не литры воды и йога"
```

Helvetii951 commented on Tracker/Modules/CreateCategory/CreateCategoryModel.swift:
```diff
@@ -0,0 +1,10 @@
+
+final class CreateCategoryModel {
+    // В рамках CreateCategory имеет ли смысл что либо сюда добавлять?
+}
```
Класс в проекте не используется. Есть смысл его совсем удалить.

DmitryBeton committed:
refactor(onboarding): единый контроллер страниц и модель данных

- Заменил OnboardingViewOne/Two на общий OnboardingPageViewController
- Добавил структуру PageModel для конфигурации страниц
- Исправил доступ к данным страницы в SceneDelegate

DmitryBeton committed:
refactor(categoryView): протокол для ViewModel и кастомная ячейка

- Добавил CategoryViewModelProtocol для инъекции зависимостей
- Создал CategoryTableViewCell с логикой конфигурации
